spec/GraphQLOverHTTP.md

```diff
@@ -237,120 +311,304 @@ omitted.
+The body of the server's response MUST follow the requirements for a
+[GraphQL response](https://graphql.github.io/graphql-spec/draft/#sec-Response),
+encoded directly in the chosen media type.
 
-A server MUST return a `Content-Type` HTTP Header with a value of a valid
-GraphQL content type. If there is no `Accept` header in the request, the
-response MUST be serialized as JSON and MUST include a
-`Content-Type: application/graphql+json` header.
+A server MUST indicate the media type of the response with a `Content-Type`
+header.
 
-If another content type is preferable to a client, it MAY include an `Accept`
-HTTP header listing other acceptable content types in order of preference. In
-this case a client SHOULD include `application/graphql+json` in the list,
-according to their preferred priority.
+If an `Accept` header is provided, the server MUST respect the given `Accept`
+header and attempt to encode the response in the highest priority media type
+listed that is supported by the server.
 
-The server MUST respect the given `Accept` header and attempt to encode the
-response in the first supported content type listed. According to the
+In alignment with the
 [HTTP 1.1 Accept](https://tools.ietf.org/html/rfc7231#section-5.3.2)
-specification, when a client does not include at least one supported content
-type in the `Accept` HTTP header, the server MAY choose to respond in one of
-several ways. The server MUST either:
+specification, when a client does not include at least one supported media type
+in the `Accept` HTTP header, the server MUST either:
+
+1. Disregard the `Accept` header and respond with the default media type of
+   `application/json`, specifying this in the `Content-Type` header; OR
+2. Respond with a `406 Not Acceptable` status code and stop processing the
+   request.
+
+A server MUST support requests which accept the `application/json` media type
+(as indicated by the `Accept` header).
+
+A server SHOULD support requests which accept the `application/graphql+json`
+media type (as indicated by the `Accept` header).
+
+Note: Prior to this specification, the media type `application/json` was in wide
+use for the HTTP response payload type. Unfortunately this means clients cannot
+trust responses from the server that do not use an HTTP 2xx status code (since
+these replies may come from non-compliant HTTP servers or proxies somewhere in
+the request chain). For this reason, this specification introduces the
+`application/graphql+json` media type on responses; however, to give existing
+servers time to move over, it is not required to be supported until 1st
+January 2025.
+
+### Legacy watershed
+
+From 1st January 2025 (`2025-01-01T00:00:00Z`), a server MUST support requests
+which accept the `application/graphql+json` media type (as indicated by the
+`Accept` header).
+
+Before 1st January 2025 (`2025-01-01T00:00:00Z`), if the client does not supply
```

Comment:
Is the implication here that spec-compatible servers should literally contain a conditional checking the current time (until we're in 2025 and that conditional can be removed)? Or I guess because this is merely a SHOULD it is OK for a server author to just decide to transition their code from one default to the other at a point that feels reasonable?

Reply:
1. Should it contain a conditional check on the time? Yes, ideally.
2. Can it just decide to transition the code from one default to another at a point that feels reasonable? Yes, knowing that it could cause issues for legacy clients.

Note: client's SHOULD include `Accept` and SHOULD accept `application/graphql+json`, so the only clients affected by this will be clients who a) do not issue `Accept` headers, and b) actually bother to read the `Content-Type` of the response and explicitly do not support `application/graphql+json`. I wonder what the size of that Venn diagram overlap is - perhaps you have data on that at Apollo?

I wonder if we should make it so that clients MUST include an `Accept` header from the watershed?

Reply:
Yeah; curl et al is the main reason that I left this as a SHOULD in the first place; but it's also why I figure if you don't supply `Accept` you probably don't care about what the server says the `Content-Type` is :laughing:

Reply:
> Should it contain a conditional check on the time? Yes, ideally.

I see this differently.  I don't see the purpose of a watershed as to set a precise date at which functionality should be turned on.  The way I see it, a watershed is to set a date at which a server without such functionality is no longer compliant.  During the time between now and the date of the watershed, new releases of a server should gradually start to support the new functionality described in the spec.  Ideally (and totally at the discretion of the implementor), a server would be backward compatible.

Reply:
I broadly agree, but David has pointed out the point where there is a sudden change in the recommended behaviour - the interpretation of a request without an `Accept` header should be `application/json` before and `application/graphql+json` after. That said, since it's only a SHOULD it doesn't matter too much and you don't _need_ a time check.